extmod/modopenamp_rproc.c

```diff
@@ -0,0 +1,161 @@
+
+// Port-defined image store operations.
+extern struct image_store_ops mp_openamp_rproc_store_ops;
+
+// Port-defined remote-proc operations.
+const struct remoteproc_ops mp_openamp_rproc_ops = {
+    .init = mp_openamp_rproc_init,
+    .mmap = mp_openamp_rproc_mmap,
+    .start = mp_openamp_rproc_start,
+    .stop = mp_openamp_rproc_stop,
+    .config = mp_openamp_rproc_config,
+    .remove = mp_openamp_rproc_remove,
+    .shutdown = mp_openamp_rproc_shutdown,
+};
+
+STATIC mp_obj_t rproc_start(mp_obj_t self_in) {
+    rproc_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    // Start the processor to run the application.
+    int error = remoteproc_start(&self->rproc);
+    if (error != 0) {
+        mp_raise_OSError(error);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(rproc_start_obj, rproc_start);
+
+STATIC mp_obj_t rproc_stop(mp_obj_t self_in) {
+    rproc_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    // Stop the processor, but the processor is not powered down.
+    int error = remoteproc_stop(&self->rproc);
+    if (error != 0) {
+        mp_raise_OSError(error);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(rproc_stop_obj, rproc_stop);
+
+STATIC mp_obj_t rproc_shutdown(mp_obj_t self_in) {
+    rproc_obj_t *self = MP_OBJ_TO_PTR(self_in);
+
+    // Shutdown the remoteproc and release its resources.
+    int error = remoteproc_shutdown(&self->rproc);
+    if (error != 0) {
+        mp_raise_OSError(error);
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(rproc_shutdown_obj, rproc_shutdown);
+
+mp_obj_t rproc_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_entry };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_entry, MP_ARG_OBJ | MP_ARG_REQUIRED,  {.u_rom_obj = MP_ROM_NONE } },
+    };
+
+    // Parse args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    rproc_obj_t *self = mp_obj_malloc_with_finaliser(rproc_obj_t, &rproc_type);
+
+    // Instantiate the remoteproc instance
+    // NOTE: ports should use rproc->priv to allocate the image store,
+    // which gets passed to remoteproc_load(), and all of the store ops.
+    remoteproc_init(&self->rproc, &mp_openamp_rproc_ops, NULL);
+
+    // Configure the remote before loading applications (optional).
+    remoteproc_config(&self->rproc, NULL);
+
+    if (mp_obj_is_int(args[ARG_entry].u_obj)) {
+        self->rproc.bootaddr = mp_obj_get_int(args[ARG_entry].u_obj);
+    } else {
+        #if MICROPY_PY_OPENAMP_RPROC_ELFLD_ENABLE
+        // Load firmware.
+        const char *path = mp_obj_str_get_str(args[ARG_entry].u_obj);
+        int error = remoteproc_load(&self->rproc, path, self->rproc.priv, &mp_openamp_rproc_store_ops, NULL);
+        if (error != 0) {
+            mp_raise_OSError(error);
+        }
+        #endif
```

Comment:
Should this have an `#else` to raise (I think) `TypeError`? (i.e. expecting an `int` because we only support the address form, although I guess you could make a case for `ValueError`?)

Reply:
For now yes, but eventually the `#else` should then try to load a `.bin` if elf loading is disabled.